src/runtimeHelpers.ts

```diff
@@ -253,6 +297,144 @@ export function unPartitionResults<T>(
+ *   [ [0, 2], [1] ],
+ *   [
+ *      [ { bar_id: 2, name: 'Burger King', rating: 3 } ],
+ *      [ { bar_id: 1, name: 'In N Out', rating: 4 }  ]
+ *   ],
+ * )
+ * ```
+ *
+ * Returns:
+ * ```
+ * [
+ *   { bar_id: 2, name: 'Burger King' }，
+ *   { bar_id: 1, rating: 4 }，
+ *   { bar_id: 2, rating: 3 }，
+ * ]
+ */
+export function unPartitionResultsByBatchKeyPartition<T>(
+    newKey: string,
+    propertyBatchKey: string,
+    batchKeyPartition: ReadonlyArray<ReadonlyArray<any>>,
+    propertyBatchKeyPartion: ReadonlyArray<ReadonlyArray<any>>,
+    /** Should be a nested array of IDs, as generated by partitionItems */
+    requestGroups: ReadonlyArray<ReadonlyArray<number>>,
+    /** The results back from the service, in the same shape as groups */
+    resultGroups: ReadonlyArray<ReadonlyArray<any | CaughtResourceError | BatchItemNotFoundError>>,
+): ReadonlyArray<T | Error> {
+    /**
+     * e.g. with our inputs, produce:
+     * ```js
+     * [
+     *   [
+     *      { order: 0, result: { bar_id: 2, name: 'Burger King' },
+     *      { order: 2, result: { bar_id: 2, rating: 3 } },
+     *   ],
+     *   [
+     *      { order: 1, result: { bar_id: 1, rating: 4 } },
+     *   ]
+     * ]
+     * ```
+     */
+
+    const zippedGroups: ReadonlyArray<ReadonlyArray<{ order: number; result: T | Error }>> = requestGroups.map(
+        (ids, i) => {
+            return ids.map((id, j) => {
+                let result = null;
+                for (const element of Object.values(resultGroups)[i]) {
+                    // case 1, error in the result.
+                    if (element instanceof CaughtResourceError) {
+                        result = element;
+                        break;
+                    }
+                    // newKey has its own key-value pair at the response top level.
+                    if (Object.values(element).includes(batchKeyPartition[i][j])) {
+                        // case 2, propertyBatchKey is returned in a nested object
+                        // { bar_id: 2, properties: {name: 'Burger King', rating: 3 }}
+                        if (propertyBatchKey in element && propertyBatchKeyPartion[i][j] in element[propertyBatchKey]) {
+                            result = element;
+                            result[propertyBatchKey] = {
+                                [propertyBatchKeyPartion[i][j]]:
+                                    element[propertyBatchKey][propertyBatchKeyPartion[i][j]],
+                            };
+                        }
+                        // case 3, propertyBatchKey is not returned in a nested object, but also at the top level.
+                        // { bar_id: 2, name: 'Burger King', rating: 3 }
+                        else {
+                            result = Object.assign(
+                                {},
+                                ...[newKey, propertyBatchKeyPartion[i][j]].map((key) => ({ [key]: element[key] })),
+                            );
```

Comment:
this syntax is really confusing to me

we usually don't need to explicitly call Object.assign for merging - we can use normal spread syntax:

```
const foo = {
  ...myObj1,
  ...myObj2,
};
```

I'm a bit more confused about what's going on here though.

```
> Object.assign({}, ...['hello', 'world'] )
{ '0': 'w', '1': 'o', '2': 'r', '3': 'l', '4': 'd' }
```

:thonk:

Could we separate out these lines or something? this is making my brain hurt a bit

Reply:
I searched how to only keep selected key-value pair in an object, https://stackoverflow.com/questions/54907549/keep-only-selected-keys-in-every-object-from-array and stackoverflow gave me this.

Reply:
I'm still not following :/

Could you give a sample demo input/output of what's happening here?

Reply:
To just keep newKey ('id') and the property ('name') from the response. So that we only return what we asked for. 

For example:
```
var data = {id: '1', name: 'John', age: 42};

var keys_to_keep = ['id', 'name']

var data_new=Object.assign({}, ...keys_to_keep.map(key => ({[key]: data[key]})));

console.log(data_new)
```
output:
```
{ id: '1', name: 'John' }
```

Reply:
does https://lodash.com/docs/4.17.15#pick work?

Reply:
the `Object.assign({}, ...` logic feels sufficiently tricky to warrant using the lodash method here (I think we're already using lodash in here anyway)